custom/shared/hooks/useSharedState.js

```diff
@@ -0,0 +1,112 @@
+import React, { useState, useEffect, useCallback, useRef } from 'react';
+import { useCallState } from '../contexts/CallProvider';
+
+// @params initialValues - initial values of the shared state.
+// @params broadcast - share the state with the other participants whenever the state changes.
+export const useSharedState = ({ initialValues = {}, broadcast = true }) => {
+  const { callObject } = useCallState();
+  const stateRef = useRef(null);
+
+  const [state, setState] = useState({
+    sharedState: initialValues,
+    setAt: undefined,
+  });
+
+  // handling the app-message event, to check if the state is being shared.
+  const handleAppMessage = useCallback(
+    event => {
+      // two types of events -
+      // 1. Request shared state (request-shared-state)
+      // 2. Set shared state (set-shared-state)
+      switch (event.data?.message?.type) {
+        // if we receive a request-shared-state message type, we check if the user has any previous state,
+        // if yes, we will send the shared-state to everyone in the call.
+        case 'request-shared-state':
+          if (!stateRef.current.setAt) return;
+          // random delay to avoid too many messages being sent at the same time from different clients.
+          const randomDelay = 200 + Math.ceil(1000 * Math.random());
+          setTimeout(() => {
+            callObject.sendAppMessage(
+              {
+                message: {
+                  type: 'set-shared-state',
+                  value: stateRef.current,
+                },
+              },
+              '*',
+            );
+          }, randomDelay);
+          break;
+        // if we receive a set-shared-state message type then, we check the state timestamp with the local one and
+        // we set the latest shared-state values into the local state.
+        case 'set-shared-state':
+          if (
+            stateRef.current.setAt &&
+            stateRef.current.setAt > event.data.message.value.setAt
```

Comment:
I don't think this check works as expected, as `event.data.message.value.setAt` is a string and `stateRef.current.setAt` _can_ be either a string or a `Date`. I'd suggest to convert both to a date, or timestamps.